src/types/inc/CodepointWidthDetector.hpp

```diff
@@ -1,37 +1,73 @@
+    //
+    // If a previous call returned false (= reached the end of the string), then on the first call with
+    // the next string, beg/len will contain the parts of the grapheme cluster that are found in that
+    // new string argument. That's true even if the two strings don't join to form a single cluster.
+    // In that case beg/len will simply be an empty string. It basically tells you
+    // "Yup, that cluster in the last string was complete after all".
+    //
+    // However, width will always be updated to represent the width of the current cluster.
+    //
+    // For instance, if the first string is a narrow emoji and the second one is U+FE0F, the first call will return
+    // the emoji with a width of 1, and the second call will return U+FE0F with a width of 2.
+    // You know these two belong together because the first call returned false.
+    // The total width is not 1+2 but rather just 2.
+    const wchar_t* beg = nullptr;
+    int len = 0;
+    // width will always be between 0 or 2.
+    int width = 0;
 
-#include "convert.hpp"
+    // If GraphemeNext/Prev return false (= reached the end of the string), they'll fill these struct
+    // members with some info so that we can check if it joins with the start of the next string argument.
+    // _state is stored ~flipped, so that we can differentiate between it being unset (0) and it being set to 0 (~0 = 255).
+    int _state = 0;
+    int _last = 0;
+};
 
-// use to measure the width of a codepoint
-class CodepointWidthDetector final
+struct CodepointWidthDetector
 {
-public:
-    CodepointWidth GetWidth(const std::wstring_view& glyph) noexcept;
-    bool IsWide(const std::wstring_view& glyph) noexcept;
-    void SetFallbackMethod(std::function<bool(const std::wstring_view&)> pfnFallback) noexcept;
-    void NotifyFontChanged() noexcept;
+    static CodepointWidthDetector& Singleton() noexcept;
 
-#ifdef UNIT_TESTING
-    friend class CodepointWidthDetectorTests;
-#endif
+    // Returns false if the end of the string has been reached.
+    bool GraphemeNext(GraphemeState& s, const std::wstring_view& str) noexcept;
+    bool GraphemePrev(GraphemeState& s, const std::wstring_view& str) noexcept;
+
+    TextMeasurementMode GetMode() const noexcept;
+    void SetFallbackMethod(std::function<bool(const std::wstring_view&)> pfnFallback) noexcept;
+    void Reset(TextMeasurementMode mode) noexcept;
 
 private:
-    uint8_t _lookupGlyphWidth(char32_t codepoint, const std::wstring_view& glyph) noexcept;
-    uint8_t _checkFallbackViaCache(char32_t codepoint, const std::wstring_view& glyph) noexcept;
+    __declspec(noinline) bool _graphemeNextWcswidth(GraphemeState& s, const wchar_t* end, const wchar_t* clusterBeg) noexcept;
+    __declspec(noinline) bool _graphemePrevWcswidth(GraphemeState& s, const wchar_t* beg, const wchar_t* clusterEnd) noexcept;
+    __declspec(noinline) bool _graphemeNextConsole(GraphemeState& s, const wchar_t* end, const wchar_t* clusterBeg) noexcept;

```

Comment:
two unconditional branches per character in the "console measurement" case - is this OK?

Reply:
I think so. If this code does get called in a hot path the branches will get predicted. The first jump will be expensive because it requires moving the parameters into place, but the next jump will be effectively cost-free.

I did benchmark it through and IIRC it was still 20% faster than our current code. I think the performance of the trie lookup helps a lot.